engine/crates/engine-v2/src/plan/collected.rs

```diff
@@ -0,0 +1,83 @@
+use schema::{DataType, FieldId, Wrapping};
+
+use crate::{
+    request::{BoundFieldId, FlatTypeCondition, SelectionSetType},
+    response::{ResponseEdge, ResponseKey},
+    utils::IdRange,
+};
+
+use super::{CollectedFieldId, CollectedSelectionSetId, ConditionalFieldId, ConditionalSelectionSetId, PlanBoundaryId};
+
+#[derive(Debug)]
+pub enum AnyCollectedSelectionSet {
+    /// Generated during planning
+    Collected(CollectedSelectionSetId),
+    Conditional(ConditionalSelectionSetId),
+    /// Generated at runtime from conditional selection sets
+    RuntimeMergedConditionals {
+        ty: SelectionSetType,
+        selection_set_ids: Vec<ConditionalSelectionSetId>,
+    },
+    RuntimeCollected(Box<RuntimeCollectedSelectionSet>),
+}
+
+/// Selection set that could not be entirely collected because of type conditions, we need to know
+/// the actual type before collecting the fields.
+#[derive(Debug, Clone)]
+pub struct ConditionalSelectionSet {
+    // needed to know where to look for __typename
+    pub ty: SelectionSetType,
+    pub maybe_boundary_id: Option<PlanBoundaryId>,
```

Comment:
A boundary is a selection set with children plans. As those will need to read data and write into it, we keep track of all response objects for that selection set (`ResponseObjectBoundaryItem`). So here the selection is the boundary. At runtime, a `RuntimeCollectedSelectionSet` may have multiple of those because the type condition made the planning split stuff.

Reply:
If it's `None` it just means there are no children plan at that level, and there's no need to track response objects.